validphys2/src/validphys/photon/compute.py

```diff
@@ -0,0 +1,289 @@
+            # self.xgrid then we don't need to reshape
+            # TODO : move the reshape inside vp-setupfit
+            # xgrid_reshape(eko, targetgrid = XGrid(self.xgrid), inputgrid = XGrid(self.xgrid))
+
+            # construct PDFs
+            pdfs = np.zeros((len(eko.rotations.inputpids), len(self.xgrid)))
+            for j, pid in enumerate(eko.rotations.inputpids):
+                if pid == 22 :
+                    pdfs[j] = photon_100GeV
+                    ph_id = j
+                if pid not in self.qcd_pdfs.flavors:
+                    continue
+                pdfs[j] = np.array(
+                    [
+                        self.qcd_pdfs.xfxQ(x, self.q_in, id, pid) / x
+                        for x in self.xgrid
+                    ]
+                )
+            
+            # Apply EKO to PDFs
+            q2 = eko.mu2grid[0]
+            with eko.operator(q2) as elem:
+                pdf_final = np.einsum("ajbk,bk", elem.operator, pdfs)
+                # error_final = np.einsum("ajbk,bk", elem.error, pdfs)
+
+        photon_Q0 = pdf_final[ph_id]
+
+        # we want x * gamma(x)
+        return self.xgrid * photon_Q0
+    
+    def produce_interpolators(self):
+        """Produce the interpolation functions to be called in compute."""
+        self.photons_array = [self.compute_photon_array(id) for id in self.replicas_id]
+        self.interpolator = [
+            interp1d(self.xgrid, photon_array, fill_value=0., kind='cubic') for photon_array in self.photons_array
+        ]
+    
+    def compute(self, xgrid):
+        """
+        Compute the photon interpolating the values of self.photon_array.
+
+        Parameters
+        ----------
+        xgrid : nd.array
+            array of x values with shape (1,xgrid,1)
+        
+        Returns
+        -------
+        photon values : nd.array
+            array of photon values with shape (1,xgrid,1)
+        """
+        return [
+            self.interpolator[id](xgrid[0,:,0])[np.newaxis,:,np.newaxis]
+            for id in range(len(self.replicas_id))
+        ]
+    
+    def integrate(self):
+        """Compute the integral of the photon on the x range."""
+        return [trapezoid(self.photons_array[id], self.xgrid) for id in range(len(self.replicas_id))]
+    
+    def generate_error_matrix(self):
+        """generate error matrix to be used for the additional errors."""
+        if not self.fiatlux_runcard["additional_errors"] :
+            return None
+        extra_set = LHAPDFSet("LUXqed17_plus_PDF4LHC15_nnlo_100", "replicas")
+        # random generator must be set previously in case of parallel replicas
+        return np.array(
+            [
+                [(extra_set.xfxQ(x, self.q_in, i, 22) - extra_set.xfxQ(x, self.q_in, 0, 22)) for i in range(101, 107+1)]
+                for x in self.xgrid
+            ]
+        ) # first index must be x, while second one must be replica index
```

Comment:
I think this can be greatly simplified by calling the underlying `lhapdf` object directly.
(the `LHAPDFSet` object is useful in the generic case, but here you know exactly the pdf and flavours you want so you don't benefit at all)

```python
qs = [self.q_in]*len(self.xgrid)
res_central = np.array(extra_set.central_members.xfxQ(22, self.xgrid, qs))

res = []
for idx_member in range(101, 108):
    tmp = np.array(extra_set.members[idx_member].xfxQ(22, self.xgrid, qs))
    res.append(tmp - res_central)

return np.stack(res, axis=1)
```

This way you skip the continuous call to the central member and also avoid the loop over x. 
I think the return shape is also correct with axis=1 (but obviously check that what I've written is equal to what you have).

The `np.array` I think are necessary because lhapdf return lists but check.